### ISSUE TITLE
exp: generate training dataset that includes pause tokens

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#34]
## Kind

document_structures

## Question

Can we improve task-level accuracy by adding `<think>` tokens at inference time (assuming the model was trained on these)?

## Hypothesis

At inference time we can use `<think>` tokens to allow the model to spend some more cycles without having to condition on additional tokens. This might give an accuracy boost.

## Background

https://arxiv.org/abs/2310.02226

## Approach

For this experiment we just want to implement the training data generation. A future experiment will run the data generation at scale, and another experiment after that will train on it.

Let's call our pause tokens `<think>`.

We want make a new document structure, `<contacts-and-distances-v2>`. This is exactly the same as `<contacts-and-distances-v1>`, except that immediately after `<begin_statements>`, with probability 0.75 we include k1 `<think> `tokens, where k1 is sampled from a geometric distribution with p=0.13 (the other 25% of the time, we include no `<think>` tokens after `<begin_statements>`). Additionally, we include max(int(k2), 0) where k2 ~ Uniform([-4, 4]) additional runs of `<think>` tokens, positioned between (never within) statements in the document.  By BETWEEN I mean that it always comes before a `<distance>` or a `<short-range-contact> <medium-range-contact> <long-range-contact>` token, so we are not breaking up statements. The length of each run of pause tokens is sampled from a geometric distribution with p=0.25 (separate lengths sampled for each run). Here is an example of the beginning of a document generated using this procedure:

```
<contacts-and-distances-v1> <begin_sequence> <ASP> <ILE> <GLN> <VAL> <GLN> <VAL> <ASN> <ILE> <ASP> <ASP> <ASN> <GLY> <LYS> <ASN> <PHE> <ASP> <TYR> <THR> <TYR> <THR> <VAL> <THR> <THR> <GLU> <SER> <GLU> <LEU> <GLN> <LYS> <VAL> <LEU> <ASN> <GLU> <LEU> <UNK> <ASP> <TYR> <ILE> <LYS> <LYS> <GLN> <GLY> <ALA> <LYS> <ARG> <VAL> <ARG> <ILE> <SER> <ILE> <THR> <ALA> <ARG> <THR> <LYS> <LYS> <GLU> <ALA> <GLU> <LYS> <PHE> <ALA> <ALA> <ILE> <LEU> <ILE> <LYS> <VAL> <PHE> <ALA> <GLU> <LEU> <GLY> <TYR> <ASN> <ASP> <ILE> <ASN> <VAL> <THR> <PHE> <ASP> <GLY> <ASP> <THR> <VAL> <THR> <VAL> <GLU> <GLY> <GLN> <LEU> <begin_statements> <think> <think> <distance> <p40> <p88> <O> <C> <d19.0> <short-range-contact> <p7> <p17> <distance> <p55> <p76> <CA> <O> <d22.0> <distance> <p47> <p83> <CZ> <C> <d16.5> <distance> <p22> <p26> <O> <O> <d8.0> <distance> <p61> <p72> <CD1> <CA> <d19.5> <short-range-contact> <p21> <p27> <think> <think> <think> <medium-range-contact> <p1> <p22>
```

When we later train on these, we will mask the loss so that <think> tokens do not count toward loss.

We still want the documents to fit in the 8192 context window and end with <end>. We should match all aspects of the <contacts-and-distances-v1> document structure.

## Success criteria

We have a new document structure: contacts-and-distances-v2 that is trained with <think> tokens as described

**Comment by timodonnell** [comment]
🤖 Opened #35 with the `contacts-and-distances-v2` training-data generator on `exp/34-contacts-and-distances-v2`.

**Scope:** generation only, as the issue requested — running at scale + training are separate follow-up issues.

**What landed:**
- `experiments/exp34_document_structures_contacts_and_distances_v2/` — `vocab.py` (v1 tokens + `<contacts-and-distances-v2>` + `<think>`, append-only so v1 ids stay stable), `parse.py`, `generate.py`, `cli.py` (`generate` + `tokenizer`), `pyproject.toml`, tests.
- Algorithm per the issue: P=0.75 gate on initial run with `k1 ~ Geom(0.13)`; `max(int(Uniform(-4,4)),0)` additional runs at random inter-statement slots with each length `~ Geom(0.25)`. Think-token overhead is subtracted from the 8192 budget before statements are allocated, so docs still end with `<end>`.
- 44 tests pass locally. Smoke run on a synthetic 60-residue chain produced an 8192-token v2 doc with 15 `<think>` tokens including a 9-token run right after `<begin_statements>`.

**Two interpretive calls** (flagged in the PR description):
1. The issue's example doc starts with `<contacts-and-distances-v1>` — treated as a paste error since the prose mandates v2.
2. At slot 0 if pLDDT also lands there, order is `<think>* <plddt_*>` (think first) to honor "immediately after `<begin_statements>`".

Happy to revise either if a different reading is intended.

---

**Summary slides:** [`plots/summary.pdf`](https://github.com/Open-Athena/MarinFold/blob/exp/34-contacts-and-distances-v2/experiments/exp34_document_structures_contacts_and_distances_v2/plots/summary.pdf) — 11 narrative slides + 9 plots on a 100-protein FoldBench-monomer sample.

Headlines:
- Docs always pack to 8192 tokens (median 8190).
- Distance statements eat **88.7%** of every doc; contacts 8.1%; `<think>` ~0.1%.
- Medium/short-range contacts almost always saturate (median fraction-captured = 1.0); **long-range is the bottleneck** (median 0.41).
- Distance count *falls* with sequence length (residue tokens crowd out the statement appendix) — worth flagging for any length-sliced downstream eval.
- `<think>` token sampling matches spec to within sampling noise (initial gate fired 81/100; additional-run counts 64/15/13/8).

Caveats are in the deck: sample is PDB monomers (not AFDB); pLDDT filter disabled because B-factors aren't pLDDT. Shapes should carry to AFDB; absolute numbers will shift.

I'll leave closing the issue to you (per the agents-never-close-experiment-issues rule). Ready for human review.
